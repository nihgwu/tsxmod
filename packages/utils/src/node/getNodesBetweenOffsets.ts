import { Node, SourceFile } from 'ts-morph'

/** Get all nodes between a start and end offset. */
export function getNodesBetweenOffsets(
  sourceFile: SourceFile,
  start: number,
  end: number
) {
  const nodes: Node[] = []
  let position = start

  while (position < end) {
    const node = sourceFile.getDescendantAtPos(position)

    if (node) {
      nodes.push(node)

      position = node.getEnd()
    } else {
      position++
    }
  }

  return nodes
}
