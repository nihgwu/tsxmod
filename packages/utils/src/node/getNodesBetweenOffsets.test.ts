import { Project } from 'ts-morph'
import { getNodesBetweenOffsets } from './getNodesBetweenOffsets'

const testSource = `
const a = 1
const b = 2
const c = 3
`.trim()

test('gets nodes within an offset range', () => {
  const project = new Project()
  const sourceFile = project.createSourceFile('test.ts', testSource)
  const nodes = getNodesBetweenOffsets(sourceFile, 11, 23)

  expect(nodes.length).toBe(4)
  expect(nodes.map((node) => node.getText())).toMatchSnapshot()
})
