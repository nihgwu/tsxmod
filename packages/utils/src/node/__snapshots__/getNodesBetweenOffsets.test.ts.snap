// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets nodes within an offset range 1`] = `
[
  "const",
  "b",
  "=",
  "2",
]
`;
